challenge-two/src/agent.ts

```diff
@@ -5,85 +5,85 @@ import {
-              tokenOut: tokenOut,
-              fee: fee.toString(),
-              recipient,
-              amountIn: amountIn.toString(),
-              amountOutMinimum: amountOutMinimum.toString(),
-              sqrtPriceLimitX96: sqrtPriceLimitX96.toString(),
-            },
-          })
+    for (const log of swapLogs) {
+      try {
+        const poolContract = new ethers.Contract(
+          log.address,
+          UNISWAP_V3_POOL_ABI,
+          provider
         );
-      } else if (swap?.name === "exactOutputSingle") {
-        const tokenIn = swap.args.params["tokenIn"];
-        const tokenOut = swap.args.params["tokenOut"];
-        const fee = swap.args.params["fee"];
-        const recipient = swap.args.params["recipient"];
-        const amountOut = swap.args.params["amountOut"];
-        const amountInMaximum = swap.args.params["amountInMaximum"];
-        const sqrtPriceLimitX96 = swap.args.params["sqrtPriceLimitX96"];
+
+        // Check if the log is from a valid Uniswap V3 pool
+        const poolFactory: string = await poolContract.callStatic.factory();
+
+        if (
+          poolFactory.toLowerCase() !== UNISWAP_V3_FACTORY_ADDRESS.toLowerCase()
+        ) {
+          // This is not a valid Uniswap V3 pool
+          continue;
+        }
+
+        // Decode the log using the ABI
+        const decodedLog = new ethers.utils.Interface(
+          UNISWAP_V3_POOL_ABI
+        ).parseLog(log);
+
+        const { sender, recipient, amount0, amount1, sqrtPriceX96, liquidity } =
+          decodedLog.args;
 
         findings.push(
           Finding.fromObject({
-            alertId: "FORTA-2",
-            name: "Nethermind Forta Bot UniSwap - exactOutputSingle",
-            description: `Swap event detected on ${swapRouter02Address}`,
+            alertId: "FORTA-3",
+            name: "Uniswap V3 Pool Swap Detected",
+            description: `Swap event detected in Uniswap V3 pool at ${log.address}`,
             severity: FindingSeverity.Low,
             type: FindingType.Info,
             metadata: {
-              tokenIn: tokenIn,
-              tokenOut: tokenOut,
-              fee: fee.toString(),
+              poolAddress: log.address,
+              sender,
               recipient,
-              amountOut: amountOut.toString(),
-              amountInMaximum: amountInMaximum.toString(),
-              sqrtPriceLimitX96: sqrtPriceLimitX96.toString(),
+              amount0: amount0.toString(),
+              amount1: amount1.toString(),
+              sqrtPriceX96: sqrtPriceX96.toString(),
+              liquidity: liquidity.toString(),
             },
           })
         );
+      } catch (error) {
+        console.error("Error processing swap log:", error);
       }
-    });
+    }
 
     return findings;
   };
 };
 
 export default {
   handleTransaction: provideHandleTransaction(
-    UNISWAP_ROUTER_ABI,
-    SWAP_ROUTER_02
+    new ethers.providers.JsonRpcProvider(RPC_PROVIDER_URL)
```

Comment:
When the bot is running in production, the correct endpoint to be used by your bot will be inject by the scanner executing it. The scanner knows which bot to run based of off the provided `chainIds` in the `package.json`. In V1, they wouldn't support some of those chains, so that why you should use the supported networks.

Additionally, you shouldn't push sensitive data to the Docker registry, such as endpoints that you're paying for. If you were trying to push public endpoints, those aren't great for performance in production.

So for the challenges, simply running the bot on the supported networks on V1 would be enough, and to rely on the endpoint provided by the scanner. This wouldn't include neither Celo nor Base.

If you need a provider in your logic, there is the `getEthersProvider` ([docs](https://docs.forta.network/en/latest/sdk/#getethersprovider))function provided by the SDK.